clearwater-config-manager.root/usr/share/clearwater/clearwater-config-manager/scripts/log_shared_config.py

```diff
@@ -0,0 +1,90 @@
+# details. You should have received a copy of the GNU General Public
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+#
+# The author can be reached by email at clearwater@metaswitch.com or by
+# post at Metaswitch Networks Ltd, 100 Church St, Enfield EN2 6BQ, UK
+#
+# Special Exception
+# Metaswitch Networks Ltd  grants you permission to copy, modify,
+# propagate, and distribute a work formed by combining OpenSSL with The
+# Software, or a work derivative of such a combination, even if such
+# copying, modification, propagation, or distribution would otherwise
+# violate the terms of the GPL. You must comply with the GPL in all
+# respects for all of the code used other than OpenSSL.
+# "OpenSSL" means OpenSSL toolkit software distributed by the OpenSSL
+# Project and licensed under the OpenSSL Licenses, or a work based on such
+# software and licensed under the OpenSSL Licenses.
+# "OpenSSL Licenses" means the OpenSSL License and Original SSLeay License
+# under which the OpenSSL Project distributes the OpenSSL toolkit software,
+# as those licenses appear in the file LICENSE-OPENSSL.
+
+import os
+import sys
+import json
+import difflib
+import subprocess
+import syslog
+
+def main():
+    '''
+    Print a readable diff of changes between etcd's shared_config and a newer
+    local copy, and log to syslog.
+    '''
+
+    # URL of shared_config etcd key
+    url = sys.argv[1]
+
+    # Get the old shared_config stored on etcd
+    # If this fails, this script is the least of our worries and we ignore, because
+    # upload_shared_config, which called this script, will print an error instead
+    jsonstr = subprocess.check_output(["curl", "-s", "-X", "GET", url])
+
+    new_config_lines = open("/etc/clearwater/shared_config").read().splitlines()
+    # etcd returns JSON; the shared_config is in node.value
+    old_config_lines = json.loads(jsonstr)["node"]["value"].splitlines()
+
+    # We're looking to log meaningful configuration changes, so sort the lines to
+    # ignore changes in line ordering
+    new_config_lines.sort()
+    old_config_lines.sort()
+    difflines = list(difflib.ndiff(old_config_lines, new_config_lines))
+
+    # (i) Pull out diff lines prefixed by "+ " / "- "; (ii) separate with quotes and commas;
+    # (iii) ignore empty lines; (iv) concatenate; (v) delete trailing comma
+    additions = ''.join(("\"" + line[2:] + "\", ") for line in difflines \
+        if line[2:] and line.startswith("+ ")).rstrip(", ")
+    deletions = ''.join(("\"" + line[2:] + "\", ") for line in difflines \
+        if line[2:] and line.startswith("- ")).rstrip(", ")
+
+    # We'll be running as root, but SUDO_USER pulls out the user who invoked sudo
+    username = os.environ['SUDO_USER']
+
+    if additions or deletions:
+        logstr = "Configuration file change: shared_config was modified by user {}. ".format(username)
+        if deletions:
+            logstr += "LINES REMOVED: "
```

Comment:
The uppercase here was so that the end of the removed lines and the beginning of the added lines could be easily spotted in a long list of changes. Usually, though, there won't be a lot of changes and I agree it doesn't look great!